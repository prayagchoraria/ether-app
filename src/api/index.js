import { ethers } from 'ethers';

export const fetchBlocks = async () => {
  const provider = new ethers.providers.Web3Provider(window.ethereum, 'any');
  const latestBlockNumber = await provider.getBlockNumber();

  const blocks = [];
  for (let i = 0; i < 10; i += 1) {
    const block = provider.getBlock(latestBlockNumber - i);
    blocks.push(block);
  }
  return Promise.all(blocks);
};
