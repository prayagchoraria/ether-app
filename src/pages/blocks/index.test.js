import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import '@testing-library/jest-dom';
import { render, waitFor } from '@testing-library/react';
import { fetchBlocks, updateBlocks } from '../../api';
import Blocks from '.';

jest.mock('../../api', () => ({
  fetchBlocks: jest.fn(),
  updateBlocks: jest.fn(),
  stopUpdating: jest.fn(),
}));

const renderBlocks = () => {
  const blocks = [
    {
      number: 12345,
      gasLimit: { _hex: '0x01ca35d2' },
      gasUsed: { _hex: '0x44f340' },
      timestamp: 1655473012,
      transactions: ['14', '2'],
    },
    {
      number: 12346,
      gasLimit: { _hex: '0x01ca35d2' },
      gasUsed: { _hex: '0x44f340' },
      timestamp: 1655473112,
      transactions: [],
    },
  ];
  fetchBlocks.mockReturnValue(
    Promise.all(blocks.map((block) => Promise.resolve(block)))
  );
  const renderedBlocks = render(
    <BrowserRouter>
      <Blocks />
    </BrowserRouter>
  );
  return renderedBlocks;
};

describe.only('Blocks Component', () => {
  it('displays the header', () => {
    const { getByText } = render(<Blocks />);
    expect(getByText('Blocks')).toBeInTheDocument();
  });

  it('renders loading state', () => {
    const { getByText } = render(<Blocks />);
    expect(getByText('Loading...')).toBeInTheDocument();
  });

  it('renders error state', async () => {
    fetchBlocks.mockReturnValue(
      Promise.reject(new Error('Something went wrong!'))
    );
    const { queryByText } = render(<Blocks />);
    expect(fetchBlocks).toHaveBeenCalledTimes(1);
    await waitFor(() => {
      expect(queryByText('Loading...')).not.toBeInTheDocument();
    });
    await waitFor(() => {
      expect(
        queryByText('Error: Something went wrong! Please refresh.')
      ).toBeInTheDocument();
    });
  });

  it('renders blocks', async () => {
    const { queryByText } = renderBlocks();
    expect(fetchBlocks).toHaveBeenCalledTimes(1);
    await waitFor(() => {
      expect(queryByText('Loading...')).not.toBeInTheDocument();
    });
    await waitFor(() => {
      expect(queryByText('12345')).toBeInTheDocument();
      expect(queryByText('12346')).toBeInTheDocument();
    });
  });

  it('matches loading snapshot', async () => {
    const tree = renderBlocks();
    expect(tree).toMatchSnapshot();
  });

  it('matches error snapshot', async () => {
    fetchBlocks.mockReturnValue(
      Promise.reject(new Error('Something went wrong!'))
    );
    const tree = render(<Blocks />);
    await waitFor(() => {
      expect(tree.queryByText('Loading...')).not.toBeInTheDocument();
    });
    expect(tree).toMatchSnapshot();
  });

  it('matches rendered snapshot', async () => {
    const tree = renderBlocks();
    await waitFor(() => {
      expect(tree.queryByText('Loading...')).not.toBeInTheDocument();
    });
    expect(tree).toMatchSnapshot();
  });

  it('updates blocks', async () => {
    const tree = renderBlocks();
    await waitFor(() => {
      expect(tree.queryByText('Loading...')).not.toBeInTheDocument();
      expect(updateBlocks).toHaveBeenCalledTimes(1);
      expect(updateBlocks).toHaveBeenCalledWith(expect.any(Function));
    });
  });
});
