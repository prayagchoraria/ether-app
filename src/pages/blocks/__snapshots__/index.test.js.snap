// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blocks Component matches error snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="blocks"
      >
        <h2>
          Blocks
        </h2>
        <span
          class="error"
        >
          Error: 
          Something went wrong! Please refresh.
        </span>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="blocks"
    >
      <h2>
        Blocks
      </h2>
      <span
        class="error"
      >
        Error: 
        Something went wrong! Please refresh.
      </span>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Blocks Component matches loading snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="blocks"
      >
        <h2>
          Blocks
        </h2>
        <span>
          Loading...
        </span>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="blocks"
    >
      <h2>
        Blocks
      </h2>
      <span>
        Loading...
      </span>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Blocks Component matches rendered snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="blocks"
      >
        <h2>
          Blocks
        </h2>
        <div
          class="data-container"
        >
          <div
            class="info-row"
          >
            <div>
              <span>
                Block
              </span>
              <span>
                12345
              </span>
            </div>
            <div>
              <span>
                Mined on
              </span>
              <span>
                17/6/2022 19:6:52
              </span>
            </div>
            <div>
              <span>
                Gas Used
              </span>
              <span>
                4518720 of 30029266
              </span>
            </div>
            <a
              class="button"
              href="/12345"
            >
              2
               transactions
            </a>
          </div>
        </div>
        <div
          class="data-container"
        >
          <div
            class="info-row"
          >
            <div>
              <span>
                Block
              </span>
              <span>
                12346
              </span>
            </div>
            <div>
              <span>
                Mined on
              </span>
              <span>
                17/6/2022 19:8:32
              </span>
            </div>
            <div>
              <span>
                Gas Used
              </span>
              <span>
                4518720 of 30029266
              </span>
            </div>
            <a
              class="button disabled"
              href="/12346"
            >
              No
               transactions
            </a>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="blocks"
    >
      <h2>
        Blocks
      </h2>
      <div
        class="data-container"
      >
        <div
          class="info-row"
        >
          <div>
            <span>
              Block
            </span>
            <span>
              12345
            </span>
          </div>
          <div>
            <span>
              Mined on
            </span>
            <span>
              17/6/2022 19:6:52
            </span>
          </div>
          <div>
            <span>
              Gas Used
            </span>
            <span>
              4518720 of 30029266
            </span>
          </div>
          <a
            class="button"
            href="/12345"
          >
            2
             transactions
          </a>
        </div>
      </div>
      <div
        class="data-container"
      >
        <div
          class="info-row"
        >
          <div>
            <span>
              Block
            </span>
            <span>
              12346
            </span>
          </div>
          <div>
            <span>
              Mined on
            </span>
            <span>
              17/6/2022 19:8:32
            </span>
          </div>
          <div>
            <span>
              Gas Used
            </span>
            <span>
              4518720 of 30029266
            </span>
          </div>
          <a
            class="button disabled"
            href="/12346"
          >
            No
             transactions
          </a>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
